echo-telemetry/src/main/java/com/netflix/spinnaker/echo/telemetry/ExecutionDataProvider.kt

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2020 Google, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.echo.telemetry
+
+import com.netflix.spinnaker.echo.api.events.Event as EchoEvent
+import com.netflix.spinnaker.echo.jackson.EchoObjectMapper
+import com.netflix.spinnaker.kork.proto.stats.CloudProvider
+import com.netflix.spinnaker.kork.proto.stats.Event as StatsEvent
+import com.netflix.spinnaker.kork.proto.stats.Execution
+import com.netflix.spinnaker.kork.proto.stats.Stage
+import com.netflix.spinnaker.kork.proto.stats.Status
+import java.util.ArrayList
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.stereotype.Component
+
+@Component
+@ConditionalOnProperty("stats.enabled")
+class ExecutionDataProvider : TelemetryEventDataProvider {
+
+  private val objectMapper = EchoObjectMapper.getInstance()
+
+  override fun populateData(echoEvent: EchoEvent, statsEvent: StatsEvent): StatsEvent {
+    val content = objectMapper.convertValue(echoEvent.getContent(), EventContent::class.java)
+    val execution = content.execution
+
+    var executionType = getExecutionType(execution.type)
+
+    if (execution.source.type.equals("templatedPipeline", ignoreCase = true)) {
```

Comment:
Wow, just learned that this works in `kotlin` even if `execution.source.type` is `null`.

Reply:
I know, it's so good!